### PR TITLE
Add proposal for 'key' modifier

### PR DESCRIPTION
**Pull request by agocke**
Implements structural equality.

### CODE REVIEW

File: proposals/key-equals.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        => EqualityContractOrigin == a.EqualityContractOrigin
+        && KeyEquals(a);
+
+    protected virtual Type EqualityContractOrigin
+        => typeof(A);
+
+    protected bool KeyEquals(A a)
+        => P1 == a.P1;
+
+    public static bool operator==(A left, A right) => left.Equals(right);
```
Comment (YairHalberstadt) [inline proposals/key-equals.md]: This doesn't check for null?

File: proposals/key-equals.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    protected bool KeyEquals(B b)
+        => base.KeyEquals(b) && P3 == a.P3;
+
+    public static bool operator==(B left, B right) => left.Equals(right);
+    public static bool operator!=(B left, B right) => !(left == right);
+}
+```
+
+
+Three synthesized members will be added to the type, `T`:
```
Comment (333fred) [inline proposals/key-equals.md]: This is well more than 3 synthesized members.

File: proposals/key-equals.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+1. override of `object.Equals(object)`. It is an error if a base class has sealed this member.
```
Comment (333fred) [inline proposals/key-equals.md]: Also consider commenting on what happens if the class has written out implementations of any of these members.

File: proposals/key-equals.md (hunk 1 above)
Comment (agocke) [reply proposals/key-equals.md]: Good point! Yeah, it seems good to make this null-aware.

File: proposals/key-equals.md (hunk 2 above)
Comment (agocke) [reply proposals/key-equals.md]: heh, just made up a number before writing this list and forgot to go back

**Comment by omariom** [comment]
`EqualityContractOrigin` is basically `GetType()`, right?

Then B.Equals could be:
```C#
public bool Equals(B other)  => other.GetType() == typeof(B) && KeyEquals(b);
```
```other.GetType() == typeof(SomeType)``` is optimised by JIT since probably .NET 2.0 and is extremely efficient. Using `EqualityContractOrigin`  means Type objects will have to be instantiated.

**Comment by CyrusNajmabadi** [comment]
> EqualityContractOrigin is basically GetType(), right?

No.  It *can* be implemented that way, but it's not required to be implemented that way.  The point is to allow the impl (including user overrides) to make the call on what determines equality.

**Comment by gafter** [comment]
> `EqualityContractOrigin` is basically `GetType()`, right?

No.  Having it be a virtual method means that you can extend a class (implementation inheritance) without changing its equality contract.

**Comment by omariom** [comment]
Can `Equals`  be implemented in a more efficient manner if **B** is sealed?
What if **B** already has its own implementation of `IEquatable<B>`.

**Comment by bugproof** [comment]
Why closed?

**Comment by agocke** [comment]
We ended up implementing records without the notion and at the moment I don't think it's worth the additional complexity.

**Pull request closed without merging**